File: geekbang/algorithm/source-code/1/lc224+227 基本计算器.cpp
// 本程序支持加、减、乘、除、括号运算
class Solution {
public:
    int calculate(string s) {
        s += " ";
        vector<string> tokens;
        string number = "";
        bool needsZero = true;
        for (char ch : s) {
            if (ch >= '0' && ch <= '9') {
                number += ch;
                needsZero = false;
                continue;
            } else {
                if (!number.empty()) {
                    tokens.push_back(number);
                    number = "";
                }
            }
            if (ch == ' ') continue;
            if (ch == '(') {
                ops.push(ch);
                needsZero = true;
                continue;
            }
            if (ch == ')') {
                while (ops.top() != '(') {
                    tokens.push_back(string(1, ops.top())); // 构造一个包含1个字符的字符串
                    ops.pop();
                }
                ops.pop();
                needsZero = false;
                continue;
            }
            if ((ch == '+' || ch == '-') && needsZero) {
                tokens.push_back("0");
            }
            int currRank = getRank(ch);
            while (!ops.empty() && getRank(ops.top()) >= currRank) {
                tokens.push_back(string(1, ops.top())); // 构造一个包含1个字符的字符串
                ops.pop();
            }
            ops.push(ch);
            needsZero = true;
        }
        while (!ops.empty()) {
            tokens.push_back(string(1, ops.top()));
            ops.pop();
        }
        return evalRPN(tokens);
/*
3+2*2
运算符栈：+ *
后缀表达式：3 2 2 * +

3+2-2
运算符栈：
后缀表达式：3 2 + 2 -
*/
    }
private:
    stack<char> ops;

    int getRank(char ch) {
        if (ch == '*' || ch == '/') return 2;
        if (ch == '+' || ch == '-') return 1;
        return 0;
    }

    long long evalRPN(vector<string>& tokens) {
        for (string& token : tokens) {
            if (token == "+" || token == "-" || token == "*" || token == "/") {
                long long y = s.top();
                s.pop();
                long long x = s.top();
                s.pop();
                long long z = calc(x, y, token);
                s.push(z);
            } else {
                s.push(atoi(token.c_str())); // 字符串转化为整数
            }
        }
        return s.top();
    }

    stack<long long> s;

    long long calc(long long x, long long y, string& op) {
        if (op == "+") return x + y;
        if (op == "-") return x - y;
        if (op == "*") return x * y;
        if (op == "/") return x / y;
        return 0;
    }
};

File: geekbang/algorithm/source-code/1/lc224+227 基本计算器.py
# 本程序支持加、减、乘、除、括号运算
class Solution:
    def calc(self, x, y, op):
        if op == "+":
            return x + y
        if op == "-":
            return x - y
        if op == "*":
            return x * y
        if op == "/":
            return int(x / y)
        return 0

    def evalRPN(self, tokens: List[str]) -> int:
        s = []
        for token in tokens:
            if token == "+" or token == "-" or token == "*" or token == "/":
                y = s[-1]
                s.pop()
                x = s[-1]
                s.pop()
                z = self.calc(x, y, token)
                s.append(z)
            else:
                s.append(int(token))
        return s[-1]

    def getRank(self, ch):
        if ch == '*' or ch == '/':
            return 2
        if ch == '+' or ch == '-':
            return 1
        return 0

    def calculate(self, s: str) -> int:
        s += " "
        tokens = []
        ops = []
        number = ""
        needsZero = True
        for ch in s:
            if ch >= '0' and ch <= '9':
                number += ch
                needsZero = False
                continue
            else:
                if len(number) > 0:
                    tokens.append(number)
                    number = ""
            if ch == ' ':
                continue
            if ch == '(':
                ops.append(ch)
                needsZero = True
                continue
            if ch == ')':
                while ops[-1] != '(':
                    tokens.append(ops[-1])
                    ops.pop()
                ops.pop()
                needsZero = False
                continue
            if (ch == '+' or ch == '-') and needsZero:
                tokens.append("0")
            currRank = self.getRank(ch)
            while len(ops) > 0 and self.getRank(ops[-1]) >= currRank:
                tokens.append(ops[-1])
                ops.pop()
            ops.append(ch)
            needsZero = True
        while len(ops) > 0:
            tokens.append(ops[-1])
            ops.pop()
        return self.evalRPN(tokens)